apt/minimization/minimizer.py

```diff
@@ -180,43 +201,110 @@ def fit(self, X=None, y=None):
+            X_test_QI = X_test.loc[:, self.quasi_identifiers_features]
 
             # collect feature data (such as min, max)
-            train_data = pd.DataFrame(X_train, columns=self._features)
             feature_data = {}
-            for feature in self._features:
-                if not feature in feature_data.keys():
-                    values = list(train_data.loc[:, feature])
+            for feature in self.quasi_identifiers_features:
+                if feature not in feature_data.keys():
                     fd = {}
-                    fd['min'] = min(values)
-                    fd['max'] = max(values)
+                    values = list(x_QI.loc[:, feature])
+                    if feature not in self.categorical_features:
+                        fd['min'] = min(values)
+                        fd['max'] = max(values)
+                        fd['range'] = max(values) - min(values)
+                    else:
+                        fd['range'] = len(values)
                     feature_data[feature] = fd
 
+            # prepare data for DT
+            categorical_features = [f for f in self._features if f in self.categorical_features and
+                                    f in self.quasi_identifiers_features]
+
+            numeric_transformer = Pipeline(
+                steps=[('imputer', SimpleImputer(strategy='constant', fill_value=0))]
+            )
+
+            # numeric_features = list(self._features) - list(self.categorical_features)
+            numeric_features = [f for f in self._features if f not in self.categorical_features and
+                                f in self.quasi_identifiers_features]
+            categorical_transformer = OneHotEncoder(handle_unknown="ignore", sparse=False)
+
+            preprocessor_QI_features = ColumnTransformer(
+                transformers=[
+                    ("num", numeric_transformer, numeric_features),
+                    ("cat", categorical_transformer, categorical_features),
+                ]
+            )
+            preprocessor_QI_features.fit(x_QI)
+
+            x_prepared = preprocessor_QI_features.transform(X_train_QI)
+
+            # preprocessor to fit data that have features not included in QI (to get accuracy)
+            numeric_features = [f for f in self._features if f not in self.categorical_features]
+            numeric_transformer = Pipeline(
+                steps=[('imputer', SimpleImputer(strategy='constant', fill_value=0))]
+            )
+            categorical_transformer = OneHotEncoder(handle_unknown="ignore", sparse=False)
+            preprocessor = ColumnTransformer(
+                transformers=[
+                    ("num", numeric_transformer, numeric_features),
+                    ("cat", categorical_transformer, self.categorical_features),
+                ]
+            )
+            preprocessor.fit(X)
+            self.preprocessor = preprocessor
```

Comment:
I know it was already like this before, but since this is an internal property I suggest to rename it to self._preprocessor